ml/ppo_encoders.py

```diff
@@ -100,32 +100,38 @@ def forward(
 
         # 3) Compute logits for existing next states (here I use dot product for scores receiving)
         # Shape: [batch_size, max_neighbors_num]
-        neighbors_logits = torch.einsum(
-            'bnd,bd->bn',
+        neighbors_logits = torch.nn.functional.cosine_similarity(
             next_state_embeddings,
-            ideal_transition_embedding
-        )
+            torch.zeros(next_state_embeddings.shape) + ideal_transition_embedding[:, None, :], dim=2
+        ) * 10
+        # neighbors_logits = torch.einsum(
+        #     'bnd,bd->bn',
+        #     next_state_embeddings,
+        #     ideal_transition_embedding
+        # )
         # TODO[Vladimir Baikalov]: Probably it's a good idea to divide logits to make the distribution smoother
-        neighbors_logits[~neighbor_node_embeddings.mask] = -torch.inf
+        inf_tensor = torch.zeros(neighbors_logits.shape)
+        inf_tensor[~neighbor_node_embeddings.mask] = -1e10
+        neighbors_logits = neighbors_logits + inf_tensor
 
         # 4) Get probs from logits
         # Shape: [batch_size, max_neighbors_num]
         neighbors_probs = torch.nn.functional.softmax(neighbors_logits, dim=1)
-        neighbors_probs[~neighbor_node_embeddings.mask] = 0  # Make sure we won't sample from padding
+        neighbors_probs = neighbors_probs * neighbor_node_embeddings.mask  # Make sure we won't sample from padding
```

Comment:
Are you sure that the mask will not change because of the optimizer step?

Reply:
Mask has no gradient?

Reply:
My bad. I forgot that the mask is re-newed and then used in forward only.